### PR TITLE
feat: auto-detect Apple MPS for inference

### PR DESCRIPTION
**Pull request by minkyu-choi07**
Tracks https://github.com/Synthefy/nori-monorepo/issues/249

## Summary
- make `NoriRegressor(device=None)` prefer CUDA, then Apple MPS, then CPU
- resolve placement once per fit and reuse it for named text encoders
- expose fitted `device_` and fail early for unavailable explicit MPS
- add macOS to the uv lock and one real Apple Silicon smoke test covering Nori + MiniLM
- budget retained MPS context caches from Metal's recommended working-set size
- keep RBF preprocessing in float32 on MPS, where float64 is unsupported

## Validation
- `uv run pytest` (403 passed, 6 skipped)
- `uv run pytest -m slow tests/test_inference_e2e.py -k "not mps"` (3 passed)
- `uv run --package synthefy pytest libs/synthefy/tests -q` (222 passed, 2 skipped)
- `uv run ruff check src scripts tests libs/synthefy/src libs/synthefy/tests ci`
- `uv build`
- `uv build --package synthefy`
- macOS arm64 lock dry-run
- real macOS arm64 MPS smoke test in CI

### CODE REVIEW

**Comment by minkyu-choi07** [comment]
Accuracy follow-up in `3cde5f8`:

- MPS inference now disables float16 autocast and runs the model in float32. Target normalization was already identical across devices; the larger accuracy drift came from full-model MPS autocast, not the small float32/float64 RBF-center rounding.
- The Apple Silicon gate now checks quality against CPU on the same fixed linear task, rather than checking only shape and finite outputs.
- Acceptance criteria: MPS correlation above `0.8`, no more than `0.02` behind CPU correlation, and MPS RMSE no more than `1.1x` CPU RMSE.

The real `macos-15` arm64 job passed under the locked PyTorch 2.8 environment: https://github.com/Synthefy/synthefy-nori/actions/runs/32064100046/job/95491899059

Local validation also passes: `401 passed, 6 skipped, 55 deselected`, Ruff clean, and both distributions build.

File: src/synthefy_nori/api.py [hunk 1]
```diff
@@ -247,23 +271,39 @@ def fit(self, X, y):
         # predict time, where an incoherent config would only surface minutes into a
         # job. The coerced policy is discarded: this call exists for its errors.
         MemoryPolicy.coerce(self.memory_policy)
+        # Resolve automatic placement once per fit so text preprocessing and model
+        # inference cannot independently choose different devices. Keep the raw
+        # constructor parameter unchanged for sklearn clone/get_params semantics.
+        self.device_ = _as_device(self.device)
+        requested_text_device = (
```
Comment (raimishah) [inline src/synthefy_nori/api.py]: why do we need this line?

File: src/synthefy_nori/api.py [hunk 2]
```diff
@@ -247,23 +271,39 @@ def fit(self, X, y):
...
             # (no embedder loaded); only text_columns=None uses the raw-array path.
             self._text_preprocessor = MultimodalPreprocessor(
                 self.text_columns, svd_dim=self.svd_dim, embedder=self.embedder,
-                device=self.device, max_cardinality=self.text_max_cardinality,
+                device=requested_text_device, max_cardinality=self.text_max_cardinality,
                 normalize=self.text_normalize)
             X_mat = self._text_preprocessor.fit_transform(X)
+            # A named encoder uses the resolved Torch device. Empty text_columns or
+            # a caller-owned callable/preloaded encoder does not place an encoder.
+            self.text_device_ = (
```
Comment (raimishah) [inline src/synthefy_nori/api.py]: same with this line? why needed?

File: src/synthefy_nori/api.py [hunk 3]
```diff
@@ -30,14 +30,36 @@
...
+        return torch.device("mps")
+    return torch.device("cpu")
 
 
 def _as_device(device):
     if device is None:
         return _default_device()
-    return torch.device(device)
+    resolved = torch.device(device)
+    if resolved.type == "mps" and not _mps_available():
```
Comment (raimishah) [inline src/synthefy_nori/api.py:66]: dont we check for mps available in _default_device function call?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -330,6 +330,29 @@ jobs:
       - run: uv sync --locked --extra dev
       - run: uv run pytest -m slow tests/test_inference_e2e.py tests/test_memory_policy_e2e.py
 
+  # Apple-Silicon gate: prove that the no-argument device path selects MPS and
+  # completes both numeric and MiniLM-backed text inference. macos-15 is an
+  # arm64 standard runner; keeping this as a real forward pass prevents mocked
+  # availability checks from advertising a backend that Nori cannot execute.
+  # Mark this job as a required status check alongside the CPU inference gate.
+  mps-inference:
```
Comment (raimishah) [inline .github/workflows/ci.yml:338]: does this actually work on our ci runners?

File: src/synthefy_nori/api.py (hunk 1 above)
Comment (minkyu-choi07) [reply src/synthefy_nori/api.py]: we don't need it - removed.

File: src/synthefy_nori/api.py (hunk 2 above)
Comment (minkyu-choi07) [reply src/synthefy_nori/api.py]: we don't need it - removed.

File: src/synthefy_nori/api.py (hunk 3 above)
Comment (minkyu-choi07) [reply src/synthefy_nori/api.py:66]: it's kind of fall back. _default_device() checks MPS only when device=None. For explicit device="mps", torch.device("mps") succeeds even if MPS is unavailable; failure occurs later during tensor/model transfer. This check provides an immediate, clear error

File: .github/workflows/ci.yml (hunk 4 above)
Comment (minkyu-choi07) [reply .github/workflows/ci.yml:338]: Yes. The hosted macos-15 job resolves to an Apple Silicon runner.

File: src/synthefy_nori/api.py (hunk 3 above)
Comment (raimishah) [reply src/synthefy_nori/api.py:66]: oh, then shouldnt we have a similar check for cuda? if they try to do cuda device and its not available?

Comment (minkyu-choi07) [reply src/synthefy_nori/api.py:66]: Yes. Added the symmetric CUDA check in 9775d9c. Explicit `cuda`/`cuda:N` now fails immediately when CUDA is unavailable, and indexed requests also fail when the index is outside `torch.cuda.device_count()`. A visible index still passes through unchanged. Coverage: unavailable, out-of-range, and valid CUDA requests; full local suite is 406 passed, 6 skipped.

**Review by raimishah — APPROVED** [review]
